server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java

```diff
@@ -239,13 +249,38 @@ private void setHighWatermark(String highWatermark) {
         this.freeDiskThresholdHigh = 100.0 - thresholdPercentageFromWatermark(highWatermark);
         this.freeBytesThresholdHigh = thresholdBytesFromWatermark(highWatermark,
             CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING.getKey());
+        setAutoReleaseStage();
     }
 
     private void setFloodStage(String floodStageRaw) {
         // Watermark is expressed in terms of used data, but we need "free" data watermark
         this.freeDiskThresholdFloodStage = 100.0 - thresholdPercentageFromWatermark(floodStageRaw);
         this.freeBytesThresholdFloodStage = thresholdBytesFromWatermark(floodStageRaw,
             CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.getKey());
+        setAutoReleaseStage();
+    }
+
+    private void setAutoReleaseIndexEnabled(boolean enabled) {
+        this.autoReleaseIndexEnabled = enabled;
+    }
+
+    /**
+     * Index Auto-Release threshold is derived from the flood stage and high watermark thresholds
+     */
+    private void setAutoReleaseStage() {
+        if (!canSetAutoRelease()) return;
+        Double freeDiskThresholdFloodStageFactor = FLOOD_STAGE_NODE_AUTO_RELEASE_FACTOR * this.freeDiskThresholdFloodStage;
+        ByteSizeValue freeBytesThresholdFloodStageFactor = new ByteSizeValue(
+            (long) FLOOD_STAGE_NODE_AUTO_RELEASE_FACTOR * this.freeBytesThresholdFloodStage.getBytes());
+
+        this.freeDiskThresholdAutoReleaseStage = freeDiskThresholdFloodStageFactor > freeDiskThresholdHigh
+            ? freeDiskThresholdFloodStageFactor : freeDiskThresholdHigh;
+        this.freeBytesThresholdAutoReleaseStage = freeBytesThresholdFloodStageFactor.getBytes() > freeBytesThresholdHigh.getBytes()
+            ? freeBytesThresholdFloodStageFactor : freeBytesThresholdHigh;
+    }
+
+    private boolean canSetAutoRelease() {
+        return (this.freeBytesThresholdHigh != null && this.freeBytesThresholdFloodStage != null);
```

Comment:
Is this needed? I can't see that `this.freeBytesThresholdHigh` is ever null, and `this.freeBytesThresholdFloodStage` should only be null during initialisation, which I think I'd prefer to fix by setting it to `new ByteSizeValue(0, ByteSizeUnit.BYTES)` first.